mdp/nodes/sfa_nodes.py

```diff
@@ -170,6 +244,9 @@ def _stop_training(self, debug=False):
         # store bias
         self._bias = mult(self.avg, self.sf)
 
+        # plot some statistics of the training data (optional)
+        if self._analysis: self._data_analysis()
```

Comment:
Same here: you have a flag to switch it off, but if somebody wants to do this step they can do it by importing your diagnostic function.

Reply:
As above: I agree this is not ideal but I don't see the benefits of a diagnostic function over a simple flag?

Reply:
Because it allows to encapsulate user code within a `try/except` without polluting the node code with diagnostics. The less flags you add ot the node, the less surprising it is to use it. Having a flag that changes the return value is considered bad practice in Python. I know that in this case the result is an Exception, but still, it seems to violate some basic code cleaness principle. If the diagnostics are happening in a different module, you have more freedom there to do whatever, without affecting existing use of the node in someone else's code.

Reply:
Right. I don't have a strong opinion here, so I'd be okay with either. However, I do feel that an additional diagnostics function is not ideal either. The basic idea behind these changes was to make the node mode accessible to other neuroscience groups, which also includes people who write less code on a daily basis.

If something breaks (which right now happens fairly often due to the eigenvalue issue) the default behavior should thus give you all the information you need to solve the problem; ideally right out of the box and without having to wrap your call int a `try` block.
I feel like there should be a cleaner way of doing this than both of the solutions proposed so far.

Reply:
Yes, and that's why a better error message is a good starting point. We could have a whole new chapter in the SFA documentation online, with code examples, about this issue. I think the problem is complicated enough that there is no "automatic" solution that would save the user from the hard job of understading why SFA is failing… We can simply help the user find her way by having better documentation.

Reply:
Absolutely! In fact, a sweet intro and tutorial on SFA is already on the list for Nils here, who's working on updating the MDP documentation :)

However, in terms of making SFA more accessible to newcomers I think we can agree that the default behavior should look like this:

* User packs data into SFA, which either successfully completes training or runs into the EV issue.
* In case of the latter **(a)** a clear error message should point out the flag to switch over to a more robust solver, and **(b)** enough information should be presented immediately to explain what is going on (maybe even a link the tutorial once it's up). Regarding the latter, Laurenz has expressed he'd like an error message to include as much information as possible, so people have a chance of understanding what the problem is and what they can do about it (both in terms of their data and what options MDP offers, i.e., optional solver-alternatives).